tests/bson/bson-document-array-access-005.phpt

```diff
@@ -0,0 +1,35 @@
+--TEST--
+MongoDB\BSON\Document array access with integers (ArrayAccess methods)
+--FILE--
+<?php
+
+require_once __DIR__ . '/../utils/basic.inc';
+
+$document = MongoDB\BSON\Document::fromPHP([
+    '0' => 'foo',
+    '1' => 'bar',
+]);
+
+// Use a variable to assert that conversion doesn't affect the original zval
+$key = 1;
```

Comment:
Interesting, how the variable could have been updated?

Reply:
Since we are operating directly on a `zval`, calling `ZVAL_NULL(key)` would update the variable, even when it wasn't passed by reference. There's no good reason to do this on purpose (despite it being used in the legacy driver), but it can happen by accident. I believe the last occurrence of us wrongly messing with the key when we were inadvertently freeing interned strings that occurred in the type map. We've gotten more careful since.

Reply:
[`Manager::__construct()`](https://github.com/mongodb/mongo-php-driver/blob/1.20.1/src/MongoDB/Manager.c#L256) is another good example of this, where we need to separate the zvals to ensure any conversions don't inadvertently change the input arguments.

This is also discussed in the [Casts and Operations](https://www.phpinternalsbook.com/php7/zvals/casts_and_operations.html) article I linked in https://github.com/mongodb/mongo-php-driver/pull/1781#discussion_r1941954106